inst/scribble/xaringanExtra-scribble.js

```diff
@@ -0,0 +1,490 @@
+/*
+ *  Scribble for remark.js/xaringan
+ *  Matthew T. Warkentin <warkentin@lunenfeld.ca>
+ */
+
+class Scribble {
+  constructor(opts) {
+    // User options
+    this.opts = opts
+    this.penColors = opts['pen_color'] || '#000000'
+    this.currColor = this.penColors[0]
+    this.penSize = opts['pen_size'] || 3
+    this.eraserSize = opts['eraser_size'] || 30
+    this.tolerance = this.eraserSize / 2
+    this.eraserColor = opts['eraser_color'] || 'rgba(0, 0, 0, 0.1)'
+    this.transparent = 'rgba(0, 0, 0, 0)'
+
+    // Fabric objects
+    this.fabrics
+    this.currFabric
+
+    // State objects
+    this.mouseDown = false
+    this.drawMode = false
+    this.eraseMode = false
+    this.hideToolbox = false
+    this.isPaused = false
+
+    this.eraserCursorMovement = this.eraserCursorMovement.bind(this)
```

Comment:
I'm curious, do you have a link where you saw this pattern? I've usually relied on renaming `self = this` inside the method that defines the event listener callback function. Which also has its drawbacks.

Reply:
I now remember why I did it this way and [here](https://stackoverflow.com/questions/11565471/removing-event-listener-which-was-added-with-bind) is the link. Since I wanted to be able to remove event listeners, I couldn't use this pattern, because `.bind()` actually creates a new function reference:
```js
foo() {
  this
}

document.addEventListener("event", this.foo.bind(this)) // no longer refers to `this.foo()`
document.removeEventListener("event", this.foo.bind(this)) // doesn't work
document.removeEventListener("event", this.foo) // doesn't work
```
But by binding outside of the event listener, you have a single function reference to refer to:
```js
constructor() {
  this.foo = this.foo.bind(this)
}

foo() {
  this
}

document.addEventListener("event", this.foo)
document.removeEventListener("event", this.foo)
```